src/Mod/Sketcher/Gui/CommandConstraints.cpp

```diff
@@ -990,6 +990,1165 @@ void CmdSketcherConstraint::activated(int /*iMsg*/)
+                || newselGeoType == Part::GeomCircle::getClassTypeId()) {
+                selCircleArc.push_back(selIdPair);
+            }
+            else if (newselGeoType == Part::GeomEllipse::getClassTypeId()
+                || newselGeoType == Part::GeomArcOfEllipse::getClassTypeId()
+                || newselGeoType == Part::GeomArcOfHyperbola::getClassTypeId()
+                || newselGeoType == Part::GeomArcOfParabola::getClassTypeId()) {
+                selEllipseAndCo.push_back(selIdPair);
+            }
+
+            bool selAllowed = makeAppropriateConstraint(onSketchPos);
+
+            if (selAllowed) {
+                // If mouse is released on something allowed, select it
+                Gui::Selection().addSelection(Obj->getDocument()->getName(),
+                    Obj->getNameInDocument(),
+                    ss.str().c_str(), onSketchPos.x, onSketchPos.y, 0.f);
+                sketchgui->draw(false, false); // Redraw
+            }
+            else {
+                if (newselGeoType == Part::GeomPoint::getClassTypeId()) {
+                    selPoints.pop_back();
+                }
+                else if (newselGeoType == Part::GeomLineSegment::getClassTypeId()) {
+                    selLine.pop_back();
+                }
+                else if (newselGeoType == Part::GeomArcOfCircle::getClassTypeId()
+                    || newselGeoType == Part::GeomCircle::getClassTypeId()) {
+                    selCircleArc.pop_back();
+                }
+                else if (newselGeoType == Part::GeomEllipse::getClassTypeId()
+                    || newselGeoType == Part::GeomArcOfEllipse::getClassTypeId()
+                    || newselGeoType == Part::GeomArcOfHyperbola::getClassTypeId()
+                    || newselGeoType == Part::GeomArcOfParabola::getClassTypeId()) {
+                    selEllipseAndCo.pop_back();
+                }
+            }
+        }
+        return true;
+    }
+protected:
+    SpecialConstraint specialConstraint;
+    AvailableConstraint availableConstraint;
+
+    Base::Vector2d previousOnSketchPos;
+
+    std::vector<SelIdPair> selPoints;
+    std::vector<SelIdPair> selLine;
+    std::vector<SelIdPair> selCircleArc;
+    std::vector<SelIdPair> selEllipseAndCo;
+
+    int numberOfConstraintsCreated;
+
+    Sketcher::SketchObject* Obj;
+
+    bool notSelectedYet(const SelIdPair& elem)
+    {
+        auto contains = [&](const std::vector<SelIdPair>& vec, const SelIdPair& elem) {
+            for (const auto& x : vec)
+            {
+                if (x.GeoId == elem.GeoId && x.PosId == elem.PosId)
+                    return true;
+            }
+            return false;
+        };
+
+        return !contains(selPoints, elem)
+            && !contains(selLine, elem)
+            && !contains(selCircleArc, elem)
+            && !contains(selEllipseAndCo, elem);
+    }
+
+    bool makeAppropriateConstraint(Base::Vector2d onSketchPos) {
```

Comment:
This method has almost 300 lines and must be refactored.

Reply:
This function is big but it is intended to be a 'map' of the constraint that you get depending on selection. It is intended to be looked at folded as such : 
![image](https://github.com/FreeCAD/FreeCAD/assets/19984177/8a29bb2f-3bce-47e8-a5e6-4308ccab7075)
Then you can open the case that you want to edit.

I don't really see how to refactor it in a way that would make more sense than this.

If you have an idea how this should be split let me know.

Reply:
Copy pasting the same code over and over makes it extremely difficult to comprehend or maintain (DRY = Don't Repeat Yourself, WET = Write Everything Twice). When the same code repeats then for sure the parameters can be placed in a simple container, perhaps constexpr, and have just one copy of the algorithm. Through the whole 300 odd lines there is IMHO much repetition that could be refactored out.

I can't remember who it was said something like "Show me code and my eyes will glaze. Show me the data and I won't need to look at the code."

Reply:
Hmm I'm still not sure how it can be refactored. Here it's convenient and easy to maintain. It's easy to find a bug. Say distance between 2 points has an issue, you go to the right section and find it easily.

I simplified it a little bit by removing unecessary tests in [00cb724](https://github.com/FreeCAD/FreeCAD/pull/9810/commits/00cb7243413ed2f514e790564116259e8980f09a)
![image](https://github.com/FreeCAD/FreeCAD/assets/19984177/b7b7f916-0186-47a9-adda-61fb601aa31b)

Reply:
> Hmm I'm still not sure how it can be refactored

On that basis I trust you are open to suggestions.

A first step could be to rationalise obvious repetitions (DRY principle) e.g.:
```
auto sp {selPoints.size()};
auto sl {selLine.size()};
auto sc {selCirCircleArc.size()};
auto se {selEllipseAndCo.size()};
```
I don't know about you, but my brain hurts if it has to read what appears to be the same thing over and over knowing subtle differences are easy to miss.

Next step, move duplicate condition code into lambdas, passing parameters. Something like
```
// reusable utility
auto matchGt = [](auto val, bool targ, bool gt){
    return gt ? val >= arg: val > targ;
};

auto checkSp = [&](bool spGt, auto spTarg, auto slTarg, auto scTarg, auto seTarg){
    return matchGt(sp, spGt, spTarg) && sl == slTarg && sc ==scTarg && se == seTarg;
}

// the work to be done
auto funcA = [](){};

if(sp > 0){
         if(checkSp(false, 1, 0, 0, 0)){ funcA(); }
    else if(checkSp(false, 2, 0, 0, 0)){ funcB(); }
    else if(checkSp(false, 1, 1, 0, 0)){ funcC(); }
    else if(checkSp(false, 3, 0, 0, 0)){ funcD(); }
    else if(checkSp(true,  4, 0, 0, 0)){ ... }
    else if(checkSp(false, 2, 1, 0, 0)){ ... }
    else if(checkSp(true,  3, 1, 0, 0)){ ... }
    else if(checkSp(true,  1, 0, 1, 0)){ ... }
    else if(checkSp(true,  1, 0, 0, 1)){ ... }
}
```
The parameters for the code, settings if you like, are being progressively moved away from the code. Code folding not required.

The conditions for checkSp can now be seen as values (set of params) which could be stored in e.g. constexpr container and looped over. More repetition is thus removed from the list of if/elses. Perhaps the loop could return an int that is used as the key of a switch to call the various funcs.

A next step could be to pass the func into the check so it is automatically called on a match. etc etc you get the idea.

With each refactor step the problem is broken into smaller and smaller chunks of reusable code. Rinse and repeat ..

As a bonus if lambdas are instead separate functions they become individually testable.

Reply:
For the first point I agree it's a easy simplification that makes a lot of sense.

For the lambda, imo this is building complexity for not much gain and it is much less flexible. First any geo type has test that are not equality. So you would need not 1 but 4 boolean. Then what if someone wants to modify and he wants to put a case where he test if se <= sl?

Regarding splitting into individual functions each case, I'm ok with it. I'm not sure it's so great because that means there's going to be 20 more functions which is a lot. But yes if they are individually testable it's probably best

Reply:
`makeAppropriateConstraint` has been refactored. @wwmayer please let me know if this is now acceptable. Thanks @berniev for the ideas

Reply:
> though 2 letters variables are going to make the CI complain

Actually, I think you'll find two-letter are ok, one-letter vars is an issue with clang-tidy.

> Thanks @berniev for the ideas

No worries. Happy to encourage more modern approaches.

> what if someone wants to modify and he wants to put a case where he test if se <= sl

Deal with what you got. YDNIY. Can be added later as reqd.

Reply:
It's in an acceptable state now. Nevertheless, one could reduce the complexity even a bit more by writing a little helper class that provides functions for the `if` conditions.
```
class GeomSelection
{
    size_t s_pts, s_lns, s_cir, s_ell;
public:
    GeomSelection(size_t s_pts, size_t s_lns, size_t s_cir, size_t s_ell)
...
   bool hasPoints() const {
       return s_pts > 0;
   }
...
   bool has1Point const {
       return s_pts == 1 && s_lns == 0 && s_cir == 0 && s_ell == 0;
   }
   bool has2Points const {
       return s_pts == 2 && s_lns == 0 && s_cir == 0 && s_ell == 0;
   }
...
};
```